File: Lab-tests/Lab2_Solutions/Q3/q3_raghav.py
# Raghav Passi

A, B = map(int, raw_input().split())
M = 10 ** 9 + 7
ans = 1
for x in bin(B)[2:][::-1]:
	if x == '1':
		ans *= A
		ans %= M
	A *= A
	A %= M
print ans

File: Lab-tests/Lab2_Solutions/Q3/q3_raghav.cpp
// Raghav Passi

#include<bits/stdc++.h>

using namespace std;

int main() {
	long long A, B, M = 1e9 + 7, ans = 1;
	cin >> A >> B;
	while(B) {
		if(B % 2) {
			ans *= A;
			ans %= M;
		}
		B /= 2;
		A *= A;
		A %= M;
	}
	cout << ans << endl;
}